{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlapping_read_write(file_gms, seq_file, out_file):\n",
    "    #work with files\n",
    "    file = open(file_gms, 'r')\n",
    "    file_reader = np.loadtxt(file, dtype ='str', skiprows=11, delimiter='\\t')\n",
    "    file.close()\n",
    "    genes_positions = []\n",
    "    for gene in file_reader:\n",
    "        genes_positions.append(gene.split(' '))\n",
    "    for gene in genes_positions:\n",
    "        if ('' in gene):\n",
    "            while('' in gene):\n",
    "                gene.remove('')\n",
    "    file_seq = open(seq_file, 'r')\n",
    "    file_reader = np.loadtxt(file_seq, dtype ='str', skiprows=1, delimiter='\\n')\n",
    "    file_seq.close()\n",
    "    #regions\n",
    "    upstream_regions = []\n",
    "    intersection = []\n",
    "    sequence = str(file_reader)\n",
    "    for gene in genes_positions:\n",
    "        s = []\n",
    "        strn = ''\n",
    "        s.append(str(gene[0]))\n",
    "        if ('<' not in gene[2]) and ('>' not in gene[3]) and (int(gene[2]) > 50) and (int(gene[3]) + 49 < len(sequence) and (int(gene[2]) + int(gene[4]) < len(sequence))):\n",
    "        \n",
    "            if gene[1] == '+':\n",
    "                for i in range(int(gene[2]) - 50, int(gene[2]) - 1):\n",
    "                    strn += str(sequence[i])\n",
    "                s.append(strn)\n",
    "                s.append(int(gene[2]) - 50)\n",
    "                s.append(int(gene[2]) - 1)\n",
    "            \n",
    "            if gene[1] == '-':\n",
    "                for i in range(int(gene[3]) - 1, int(gene[3]) + 49):\n",
    "                    strn += str(sequence[i])\n",
    "                s.append(strn[::-1])\n",
    "                s.append(int(gene[3]) - 1)\n",
    "                s.append(int(gene[3]) + 49)\n",
    "            upstream_regions.append(s)\n",
    "        elif ('<' not in gene[2]) and ('>' not in gene[3]) : \n",
    "            s.append(gene[2])\n",
    "            s.append(gene[3])\n",
    "            upstream_regions.append(s)\n",
    "    \n",
    "   # print(upstream_regions)\n",
    "    upstream_regions_complement = []\n",
    "    for i in range(len(upstream_regions)):\n",
    "        upstream_region = ''\n",
    "        for k in upstream_regions[i][1]:\n",
    "            for j in k:\n",
    "                if (j == 'T'):\n",
    "                    upstream_region += 'A'\n",
    "                if (j == 'A'):\n",
    "                    upstream_region += 'T'\n",
    "                if (j == 'C'):\n",
    "                    upstream_region += 'G'\n",
    "                if (j == 'G'):\n",
    "                    upstream_region += 'C'\n",
    "        upstream_regions_complement.append(upstream_region)\n",
    "    #overlapping\n",
    "    overlapping = []\n",
    "    for i in upstream_regions:\n",
    "        for j in genes_positions:\n",
    "            if ('<' not in j[2]) and ('>' not in j[3]):\n",
    "                if (len(j) == 9 and len(i) == 4):\n",
    "                  #  print(i, j)\n",
    "                    if (int(i[3]) > int(j[2]) and i != j and int(i[3]) < int(j[3])):\n",
    "                        s = [i[0], j[0]]\n",
    "                        k = [j[0], i[0]]\n",
    "                        if (s not in overlapping and k not in overlapping):\n",
    "                            overlapping.append(s)\n",
    "    #output_writing\n",
    "    out = open(out_file, 'w')\n",
    "    out.write('First 100 nucleotides from the upstream regions from each gene:')\n",
    "    out.write('\\n')\n",
    "    out.write('\\n')\n",
    "    for i in range(len(upstream_regions)):\n",
    "        if (genes_positions[i][5] != 'atypical'):\n",
    "            out.write('gene name: <NC_017968>-<GMS2 ')\n",
    "            out.write(str(upstream_regions[i][0]) + '>')\n",
    "            out.write('\\n')\n",
    "            out.write(str(upstream_regions[i][1]))\n",
    "            out.write('\\n')\n",
    "            out.write(str(upstream_regions_complement[i]))\n",
    "            out.write('\\n')\n",
    "            out.write('\\n')\n",
    "        else:\n",
    "            out.write('gene name: <NC_017968>-<GMS2 ')\n",
    "            out.write(str(upstream_regions[i][0]) + '>')\n",
    "            out.write('\\n')\n",
    "            out.write('NO UPSTREAM REGION FOUND')\n",
    "            out.write('\\n')\n",
    "            out.write('\\n')\n",
    "    out.write('Numbers of overlapping genes and upstream regions:')\n",
    "    out.write('\\n')\n",
    "    for i in range(len(overlapping)):\n",
    "        out.write(str(overlapping[i]))\n",
    "        out.write('\\n')\n",
    "    out.close()\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "for root, dirs, files in os.walk('/home/emil/Desktop/Science/gms2_viruses_output'):\n",
    "    for file in files:\n",
    "        path = os.path.join(root, file)\n",
    "        s = path.split('/')\n",
    "        name = s[len(s)-1].replace('_gms2.out','')\n",
    "        sequence = '/home/emil/Desktop/Science/viruses_genomes/' + name\n",
    "        outname = '/home/emil/Desktop/Science/overlapping_regions/' + name +'.overlapping.out'\n",
    "        #print(path)\n",
    "        overlapping_read_write(path, sequence, outname)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
